src/plugin.ts

```diff
@@ -152,36 +58,115 @@ function getTSConfigFile(tsconfigPath: string): ts.ParsedCommandLine {
 }
 
 /** Create a glob matching function. */
-const matchGlob = (globs: string[]) => {
-  const matchers = globs.map((g) => matcher(g, { dot: true }));
+const matchGlob = (globs?: string[]) => {
+  const matchers = (globs || []).map((g) => matcher(g, { dot: true }));
+
   return (filename: string) =>
     Boolean(filename && matchers.find((match) => match(filename)));
 };
 
 /** Inject typescript docgen information into modules at the end of a build */
-export default class DocgenPlugin {
+export default class DocgenPlugin implements WebpackPluginInstance {
   private name = "React Docgen Typescript Plugin";
   private options: PluginOptions;
 
   constructor(options: PluginOptions = {}) {
     this.options = options;
   }
 
-  apply(compiler: webpack.Compiler): void {
+  apply(compiler: Compiler): void {
+    const pluginName = "DocGenPlugin";
+    const {
+      docgenOptions,
+      compilerOptions,
+      generateOptions,
+    } = this.getOptions();
+    const docGenParser = docGen.withCompilerOptions(
+      compilerOptions,
+      docgenOptions
+    );
+    const { exclude = [], include = ["**/**.tsx"] } = this.options;
+    const isExcluded = matchGlob(exclude);
+    const isIncluded = matchGlob(include);
+
+    compiler.hooks.compilation.tap(
+      pluginName,
+      (compilation, { normalModuleFactory }) => {
+        compilation.dependencyTemplates.set(
+          // eslint-disable-next-line
+          // @ts-ignore TODO: Figure out why this isn't allowed
+          DocGenDependency,
+          new DocGenDependency.Template()
+        );
+
+        const handler = (parser: JavascriptParser) => {
+          parser.hooks.program.tap(pluginName, () => {
```

Comment:
As I understand, we can use `seal` hook as and before. It will simplify logic a bit

Reply:
Great, I'll look that up. 👍